File: aws/ses/typescript/src/types/requests.ts
/**
 * AWS SES Request Types
 *
 * This module contains type definitions for all API request types in AWS SES v2.
 */

import type {
  EmailContent,
  Destination,
  MessageTag,
  BulkEmailEntry,
  RawMessage,
  Message
} from './email.js';
import type {
  DkimSigningAttributes,
  Tag,
  IdentityType
} from './identity.js';
import type {
  ConfigurationSet,
  EventDestination,
  DeliveryOptions,
  ReputationOptions,
  SendingOptions,
  SuppressionOptions,
  TrackingOptions,
  VdmOptions
} from './configuration.js';
import type { SuppressionReason } from './suppression.js';
import type {
  Contact,
  Topic,
  TopicPreference,
  ListContactsFilter
} from './contacts.js';

/**
 * Request to send an email.
 */
export interface SendEmailRequest {
  /** The email address to send from */
  fromEmailAddress?: string;
  /** The ARN of the email identity to use */
  fromEmailAddressIdentityArn?: string;
  /** The destination for the email */
  destination?: Destination;
  /** The reply-to addresses */
  replyToAddresses?: string[];
  /** The address to receive feedback notifications */
  feedbackForwardingEmailAddress?: string;
  /** The ARN of the feedback forwarding identity */
  feedbackForwardingEmailAddressIdentityArn?: string;
  /** The email content */
  content: EmailContent;
  /** Message tags for categorization */
  emailTags?: MessageTag[];
  /** The configuration set to use */
  configurationSetName?: string;
  /** List management options */
  listManagementOptions?: ListManagementOptions;
}

/**
 * List management options for emails.
 */
export interface ListManagementOptions {
  /** The name of the contact list */
  contactListName: string;
  /** The name of the topic */
  topicName?: string;
}

/**
 * Request to send a raw email.
 */
export interface SendRawEmailRequest {
  /** The email address to send from */
  fromEmailAddress?: string;
  /** The ARN of the email identity to use */
  fromEmailAddressIdentityArn?: string;
  /** The destination addresses */
  destinations?: string[];
  /** The raw message data */
  rawMessage: RawMessage;
  /** The configuration set to use */
  configurationSetName?: string;
  /** Message tags for categorization */
  emailTags?: MessageTag[];
  /** The address to receive feedback notifications */
  feedbackForwardingEmailAddress?: string;
  /** The ARN of the feedback forwarding identity */
  feedbackForwardingEmailAddressIdentityArn?: string;
}

/**
 * Request to send bulk emails.
 */
export interface SendBulkEmailRequest {
  /** The email address to send from */
  fromEmailAddress?: string;
  /** The ARN of the email identity to use */
  fromEmailAddressIdentityArn?: string;
  /** The reply-to addresses */
  replyToAddresses?: string[];
  /** The address to receive feedback notifications */
  feedbackForwardingEmailAddress?: string;
  /** The ARN of the feedback forwarding identity */
  feedbackForwardingEmailAddressIdentityArn?: string;
  /** Default email tags to apply to all messages */
  defaultEmailTags?: MessageTag[];
  /** The default email content */
  defaultContent: EmailContent;
  /** The bulk email entries */
  bulkEmailEntries: BulkEmailEntry[];
  /** The configuration set to use */
  configurationSetName?: string;
}

/**
 * Request to create an email identity.
 */
export interface CreateEmailIdentityRequest {
  /** The email address or domain to verify */
  emailIdentity: string;
  /** Tags to associate with the identity */
  tags?: Tag[];
  /** DKIM signing attributes */
  dkimSigningAttributes?: DkimSigningAttributes;
  /** The configuration set to associate with the identity */
  configurationSetName?: string;
}

/**
 * Request to delete an email identity.
 */
export interface DeleteEmailIdentityRequest {
  /** The email address or domain to delete */
  emailIdentity: string;
}

/**
 * Request to get email identity details.
 */
export interface GetEmailIdentityRequest {
  /** The email address or domain to retrieve */
  emailIdentity: string;
}

/**
 * Request to list email identities.
 */
export interface ListEmailIdentitiesRequest {
  /** Token for pagination */
  nextToken?: string;
  /** Maximum number of results to return */
  pageSize?: number;
}

/**
 * Request to put email identity DKIM attributes.
 */
export interface PutEmailIdentityDkimAttributesRequest {
  /** The email address or domain */
  emailIdentity: string;
  /** Whether to enable DKIM signing */
  signingEnabled?: boolean;
}

/**
 * Request to put email identity DKIM signing attributes.
 */
export interface PutEmailIdentityDkimSigningAttributesRequest {
  /** The email address or domain */
  emailIdentity: string;
  /** The DKIM signing attributes */
  signingAttributes?: DkimSigningAttributes;
  /** The signing attributes origin */
  signingAttributesOrigin: 'AWS_SES' | 'EXTERNAL';
}

/**
 * Request to put email identity feedback attributes.
 */
export interface PutEmailIdentityFeedbackAttributesRequest {
  /** The email address or domain */
  emailIdentity: string;
  /** Whether email feedback forwarding is enabled */
  emailForwardingEnabled?: boolean;
}

/**
 * Request to put email identity mail from attributes.
 */
export interface PutEmailIdentityMailFromAttributesRequest {
  /** The email address or domain */
  emailIdentity: string;
  /** The custom MAIL FROM domain */
  mailFromDomain?: string;
  /** Behavior when MX record is not found */
  behaviorOnMxFailure?: 'USE_DEFAULT_VALUE' | 'REJECT_MESSAGE';
}

/**
 * Request to put email identity configuration set attributes.
 */
export interface PutEmailIdentityConfigurationSetAttributesRequest {
  /** The email address or domain */
  emailIdentity: string;
  /** The configuration set name */
  configurationSetName?: string;
}

/**
 * Request to create a configuration set.
 */
export interface CreateConfigurationSetRequest {
  /** The name of the configuration set */
  configurationSetName: string;
  /** Tracking options */
  trackingOptions?: TrackingOptions;
  /** Delivery options */
  deliveryOptions?: DeliveryOptions;
  /** Reputation options */
  reputationOptions?: ReputationOptions;
  /** Sending options */
  sendingOptions?: SendingOptions;
  /** Suppression options */
  suppressionOptions?: SuppressionOptions;
  /** VDM options */
  vdmOptions?: VdmOptions;
  /** Tags to associate with the configuration set */
  tags?: Tag[];
}

/**
 * Request to delete a configuration set.
 */
export interface DeleteConfigurationSetRequest {
  /** The name of the configuration set */
  configurationSetName: string;
}

/**
 * Request to get configuration set details.
 */
export interface GetConfigurationSetRequest {
  /** The name of the configuration set */
  configurationSetName: string;
}

/**
 * Request to list configuration sets.
 */
export interface ListConfigurationSetsRequest {
  /** Token for pagination */
  nextToken?: string;
  /** Maximum number of results to return */
  pageSize?: number;
}

/**
 * Request to put configuration set delivery options.
 */
export interface PutConfigurationSetDeliveryOptionsRequest {
  /** The name of the configuration set */
  configurationSetName: string;
  /** The TLS policy */
  tlsPolicy?: 'REQUIRE' | 'OPTIONAL';
  /** The sending pool name */
  sendingPoolName?: string;
}

/**
 * Request to put configuration set reputation options.
 */
export interface PutConfigurationSetReputationOptionsRequest {
  /** The name of the configuration set */
  configurationSetName: string;
  /** Whether reputation metrics are enabled */
  reputationMetricsEnabled?: boolean;
}

/**
 * Request to put configuration set sending options.
 */
export interface PutConfigurationSetSendingOptionsRequest {
  /** The name of the configuration set */
  configurationSetName: string;
  /** Whether sending is enabled */
  sendingEnabled?: boolean;
}

/**
 * Request to put configuration set suppression options.
 */
export interface PutConfigurationSetSuppressionOptionsRequest {
  /** The name of the configuration set */
  configurationSetName: string;
  /** List of suppression reasons */
  suppressedReasons?: SuppressionReason[];
}

/**
 * Request to put configuration set tracking options.
 */
export interface PutConfigurationSetTrackingOptionsRequest {
  /** The name of the configuration set */
  configurationSetName: string;
  /** The custom redirect domain */
  customRedirectDomain?: string;
}

/**
 * Request to create a configuration set event destination.
 */
export interface CreateConfigurationSetEventDestinationRequest {
  /** The name of the configuration set */
  configurationSetName: string;
  /** The event destination name */
  eventDestinationName: string;
  /** The event destination configuration */
  eventDestination: EventDestination;
}

/**
 * Request to update a configuration set event destination.
 */
export interface UpdateConfigurationSetEventDestinationRequest {
  /** The name of the configuration set */
  configurationSetName: string;
  /** The event destination name */
  eventDestinationName: string;
  /** The event destination configuration */
  eventDestination: EventDestination;
}

/**
 * Request to delete a configuration set event destination.
 */
export interface DeleteConfigurationSetEventDestinationRequest {
  /** The name of the configuration set */
  configurationSetName: string;
  /** The event destination name */
  eventDestinationName: string;
}

/**
 * Request to get configuration set event destinations.
 */
export interface GetConfigurationSetEventDestinationsRequest {
  /** The name of the configuration set */
  configurationSetName: string;
}

/**
 * Request to put a suppressed destination.
 */
export interface PutSuppressedDestinationRequest {
  /** The email address to suppress */
  emailAddress: string;
  /** The reason for suppression */
  reason: SuppressionReason;
}

/**
 * Request to delete a suppressed destination.
 */
export interface DeleteSuppressedDestinationRequest {
  /** The email address to remove from suppression list */
  emailAddress: string;
}

/**
 * Request to get a suppressed destination.
 */
export interface GetSuppressedDestinationRequest {
  /** The email address to retrieve */
  emailAddress: string;
}

/**
 * Request to list suppressed destinations.
 */
export interface ListSuppressedDestinationsRequest {
  /** Filter by suppression reason */
  reasons?: SuppressionReason[];
  /** Filter by start time */
  startDate?: Date;
  /** Filter by end time */
  endDate?: Date;
  /** Token for pagination */
  nextToken?: string;
  /** Maximum number of results to return */
  pageSize?: number;
}

/**
 * Request to create a contact list.
 */
export interface CreateContactListRequest {
  /** The name of the contact list */
  contactListName: string;
  /** Description of the contact list */
  description?: string;
  /** Topics associated with the contact list */
  topics?: Topic[];
  /** Tags to associate with the contact list */
  tags?: Tag[];
}

/**
 * Request to delete a contact list.
 */
export interface DeleteContactListRequest {
  /** The name of the contact list to delete */
  contactListName: string;
}

/**
 * Request to get a contact list.
 */
export interface GetContactListRequest {
  /** The name of the contact list */
  contactListName: string;
}

/**
 * Request to list contact lists.
 */
export interface ListContactListsRequest {
  /** Maximum number of results to return */
  pageSize?: number;
  /** Token for pagination */
  nextToken?: string;
}

/**
 * Request to update a contact list.
 */
export interface UpdateContactListRequest {
  /** The name of the contact list */
  contactListName: string;
  /** Topics to associate with the contact list */
  topics?: Topic[];
  /** Description of the contact list */
  description?: string;
}

/**
 * Request to create a contact.
 */
export interface CreateContactRequest {
  /** The name of the contact list */
  contactListName: string;
  /** The contact's email address */
  emailAddress: string;
  /** Topic preferences for the contact */
  topicPreferences?: TopicPreference[];
  /** Whether to unsubscribe from all topics */
  unsubscribeAll?: boolean;
  /** Additional attributes as a JSON string */
  attributesData?: string;
}

/**
 * Request to delete a contact.
 */
export interface DeleteContactRequest {
  /** The name of the contact list */
  contactListName: string;
  /** The contact's email address */
  emailAddress: string;
}

/**
 * Request to get a contact.
 */
export interface GetContactRequest {
  /** The name of the contact list */
  contactListName: string;
  /** The contact's email address */
  emailAddress: string;
}

/**
 * Request to list contacts.
 */
export interface ListContactsRequest {
  /** The name of the contact list */
  contactListName: string;
  /** Filter criteria */
  filter?: ListContactsFilter;
  /** Maximum number of results to return */
  pageSize?: number;
  /** Token for pagination */
  nextToken?: string;
}

/**
 * Request to update a contact.
 */
export interface UpdateContactRequest {
  /** The name of the contact list */
  contactListName: string;
  /** The contact's email address */
  emailAddress: string;
  /** Topic preferences for the contact */
  topicPreferences?: TopicPreference[];
  /** Whether to unsubscribe from all topics */
  unsubscribeAll?: boolean;
  /** Additional attributes as a JSON string */
  attributesData?: string;
}

/**
 * Request to get account details.
 */
export interface GetAccountRequest {
  // No parameters needed
}

/**
 * Request to put account details.
 */
export interface PutAccountDetailsRequest {
  /** The type of email your account will send */
  mailType?: MailType;
  /** The URL of your website */
  websiteURL?: string;
  /** The language you want to use for contacting AWS */
  contactLanguage?: ContactLanguage;
  /** The use case description */
  useCaseDescription?: string;
  /** Additional email addresses to contact */
  additionalContactEmailAddresses?: string[];
  /** Whether to review the latest decision */
  productionAccessEnabled?: boolean;
}

/**
 * The type of email being sent.
 */
export type MailType = 'MARKETING' | 'TRANSACTIONAL';

/**
 * The language for AWS contact.
 */
export type ContactLanguage = 'EN' | 'JA';

/**
 * Request to put account suppression attributes.
 */
export interface PutAccountSuppressionAttributesRequest {
  /** List of suppression reasons to enable at account level */
  suppressedReasons?: SuppressionReason[];
}

/**
 * Request to get account suppression attributes.
 */
export interface GetAccountSuppressionAttributesRequest {
  // No parameters needed
}

/**
 * Request to put account sending attributes.
 */
export interface PutAccountSendingAttributesRequest {
  /** Whether sending is enabled for the account */
  sendingEnabled?: boolean;
}

/**
 * Request to put account VDM attributes.
 */
export interface PutAccountVdmAttributesRequest {
  /** VDM attributes for the account */
  vdmAttributes: VdmOptions;
}

/**
 * Request to get account VDM attributes.
 */
export interface GetAccountVdmAttributesRequest {
  // No parameters needed
}

/**
 * Request to create an email template.
 */
export interface CreateEmailTemplateRequest {
  /** The name of the template */
  templateName: string;
  /** The content of the template */
  templateContent: EmailTemplateContent;
}

/**
 * Email template content.
 */
export interface EmailTemplateContent {
  /** The subject line of the template */
  subject?: string;
  /** The HTML body of the template */
  html?: string;
  /** The plain text body of the template */
  text?: string;
}

/**
 * Request to update an email template.
 */
export interface UpdateEmailTemplateRequest {
  /** The name of the template */
  templateName: string;
  /** The content of the template */
  templateContent: EmailTemplateContent;
}

/**
 * Request to delete an email template.
 */
export interface DeleteEmailTemplateRequest {
  /** The name of the template to delete */
  templateName: string;
}

/**
 * Request to get an email template.
 */
export interface GetEmailTemplateRequest {
  /** The name of the template */
  templateName: string;
}

/**
 * Request to list email templates.
 */
export interface ListEmailTemplatesRequest {
  /** Token for pagination */
  nextToken?: string;
  /** Maximum number of results to return */
  pageSize?: number;
}

/**
 * Request to test render an email template.
 */
export interface TestRenderEmailTemplateRequest {
  /** The name of the template */
  templateName: string;
  /** The template data as a JSON string */
  templateData: string;
}

File: aws/ses/rust/src/types/requests.rs
//! API request types for SES v2.

use serde::{Deserialize, Serialize};
use super::{
    BulkEmailEntry, Destination, DkimSigningAttributes,
    EmailContent, EventDestination, MailFromAttributes, MessageTag,
    SuppressionListReason, Topic,
};

/// Request to send an email.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct SendEmailRequest {
    /// The email address to use as the "From" address.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub from_email_address: Option<String>,
    /// An alternative "From" address to use (for authorized senders).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub from_email_address_identity_arn: Option<String>,
    /// Destination (recipients).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub destination: Option<Destination>,
    /// Reply-to addresses.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reply_to_addresses: Option<Vec<String>>,
    /// Email address for bounces and complaints.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub feedback_forwarding_email_address: Option<String>,
    /// Email content.
    pub content: EmailContent,
    /// Message tags.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email_tags: Option<Vec<MessageTag>>,
    /// Configuration set name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub configuration_set_name: Option<String>,
    /// List management options.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub list_management_options: Option<ListManagementOptions>,
}

impl SendEmailRequest {
    /// Create a new send email request.
    pub fn new(content: EmailContent) -> Self {
        Self {
            from_email_address: None,
            from_email_address_identity_arn: None,
            destination: None,
            reply_to_addresses: None,
            feedback_forwarding_email_address: None,
            content,
            email_tags: None,
            configuration_set_name: None,
            list_management_options: None,
        }
    }

    /// Set the from address.
    pub fn with_from(mut self, from: impl Into<String>) -> Self {
        self.from_email_address = Some(from.into());
        self
    }

    /// Set destination.
    pub fn with_destination(mut self, destination: Destination) -> Self {
        self.destination = Some(destination);
        self
    }

    /// Add a reply-to address.
    pub fn add_reply_to(mut self, reply_to: impl Into<String>) -> Self {
        self.reply_to_addresses
            .get_or_insert_with(Vec::new)
            .push(reply_to.into());
        self
    }

    /// Set configuration set.
    pub fn with_configuration_set(mut self, name: impl Into<String>) -> Self {
        self.configuration_set_name = Some(name.into());
        self
    }

    /// Add a message tag.
    pub fn add_tag(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
        self.email_tags
            .get_or_insert_with(Vec::new)
            .push(MessageTag::new(name, value));
        self
    }
}

/// Request to send bulk email.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct SendBulkEmailRequest {
    /// The email address to use as the "From" address.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub from_email_address: Option<String>,
    /// An alternative "From" address to use.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub from_email_address_identity_arn: Option<String>,
    /// Reply-to addresses.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reply_to_addresses: Option<Vec<String>>,
    /// Feedback forwarding email address.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub feedback_forwarding_email_address: Option<String>,
    /// Default email content.
    pub default_content: EmailContent,
    /// Bulk email entries.
    pub bulk_email_entries: Vec<BulkEmailEntry>,
    /// Configuration set name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub configuration_set_name: Option<String>,
    /// Default email tags.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_email_tags: Option<Vec<MessageTag>>,
}

impl SendBulkEmailRequest {
    /// Create a new bulk email request.
    pub fn new(default_content: EmailContent, entries: Vec<BulkEmailEntry>) -> Self {
        Self {
            from_email_address: None,
            from_email_address_identity_arn: None,
            reply_to_addresses: None,
            feedback_forwarding_email_address: None,
            default_content,
            bulk_email_entries: entries,
            configuration_set_name: None,
            default_email_tags: None,
        }
    }

    /// Set from address.
    pub fn with_from(mut self, from: impl Into<String>) -> Self {
        self.from_email_address = Some(from.into());
        self
    }

    /// Set configuration set.
    pub fn with_configuration_set(mut self, name: impl Into<String>) -> Self {
        self.configuration_set_name = Some(name.into());
        self
    }
}

/// List management options for email.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct ListManagementOptions {
    /// The name of the contact list.
    pub contact_list_name: String,
    /// The name of the topic.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub topic_name: Option<String>,
}

/// Request to create an email identity.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct CreateEmailIdentityRequest {
    /// The email address or domain to verify.
    pub email_identity: String,
    /// Tags to associate with the identity.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<super::Tag>>,
    /// DKIM signing attributes.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dkim_signing_attributes: Option<DkimSigningAttributes>,
    /// Configuration set name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub configuration_set_name: Option<String>,
}

impl CreateEmailIdentityRequest {
    /// Create a new email identity request.
    pub fn new(email_identity: impl Into<String>) -> Self {
        Self {
            email_identity: email_identity.into(),
            tags: None,
            dkim_signing_attributes: None,
            configuration_set_name: None,
        }
    }

    /// Add a tag.
    pub fn add_tag(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.tags
            .get_or_insert_with(Vec::new)
            .push(super::Tag::new(key, value));
        self
    }

    /// Set configuration set.
    pub fn with_configuration_set(mut self, name: impl Into<String>) -> Self {
        self.configuration_set_name = Some(name.into());
        self
    }
}

/// Request to delete an email identity.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct DeleteEmailIdentityRequest {
    /// The email identity to delete.
    pub email_identity: String,
}

/// Request to create a configuration set.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct CreateConfigurationSetRequest {
    /// The name of the configuration set.
    pub configuration_set_name: String,
    /// Tracking options.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tracking_options: Option<super::TrackingOptions>,
    /// Delivery options.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub delivery_options: Option<super::DeliveryOptions>,
    /// Reputation options.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reputation_options: Option<super::ReputationOptions>,
    /// Sending options.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sending_options: Option<super::SendingOptions>,
    /// Tags.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<super::Tag>>,
    /// Suppression options.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub suppression_options: Option<super::SuppressionOptions>,
}

impl CreateConfigurationSetRequest {
    /// Create a new configuration set request.
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            configuration_set_name: name.into(),
            tracking_options: None,
            delivery_options: None,
            reputation_options: None,
            sending_options: None,
            tags: None,
            suppression_options: None,
        }
    }
}

/// Request to delete a configuration set.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct DeleteConfigurationSetRequest {
    /// The name of the configuration set.
    pub configuration_set_name: String,
}

/// Request to create an email template.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct CreateEmailTemplateRequest {
    /// The name of the template.
    pub template_name: String,
    /// Template content.
    pub template_content: TemplateContent,
}

impl CreateEmailTemplateRequest {
    /// Create a new template request.
    pub fn new(name: impl Into<String>, content: TemplateContent) -> Self {
        Self {
            template_name: name.into(),
            template_content: content,
        }
    }
}

/// Template content.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct TemplateContent {
    /// The subject line of the email.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subject: Option<String>,
    /// The HTML body of the email.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub html: Option<String>,
    /// The plain text body of the email.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
}

impl TemplateContent {
    /// Create new template content.
    pub fn new() -> Self {
        Self {
            subject: None,
            html: None,
            text: None,
        }
    }

    /// Set subject.
    pub fn with_subject(mut self, subject: impl Into<String>) -> Self {
        self.subject = Some(subject.into());
        self
    }

    /// Set HTML body.
    pub fn with_html(mut self, html: impl Into<String>) -> Self {
        self.html = Some(html.into());
        self
    }

    /// Set text body.
    pub fn with_text(mut self, text: impl Into<String>) -> Self {
        self.text = Some(text.into());
        self
    }
}

impl Default for TemplateContent {
    fn default() -> Self {
        Self::new()
    }
}

/// Request to update an email template.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct UpdateEmailTemplateRequest {
    /// The name of the template.
    pub template_name: String,
    /// Template content.
    pub template_content: TemplateContent,
}

/// Request to delete an email template.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct DeleteEmailTemplateRequest {
    /// The name of the template.
    pub template_name: String,
}

/// Request to create a contact list.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct CreateContactListRequest {
    /// The name of the contact list.
    pub contact_list_name: String,
    /// Topics for the contact list.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub topics: Option<Vec<Topic>>,
    /// Description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Tags.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<super::Tag>>,
}

impl CreateContactListRequest {
    /// Create a new contact list request.
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            contact_list_name: name.into(),
            topics: None,
            description: None,
            tags: None,
        }
    }

    /// Set description.
    pub fn with_description(mut self, description: impl Into<String>) -> Self {
        self.description = Some(description.into());
        self
    }

    /// Set topics.
    pub fn with_topics(mut self, topics: Vec<Topic>) -> Self {
        self.topics = Some(topics);
        self
    }
}

/// Request to delete a contact list.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct DeleteContactListRequest {
    /// The name of the contact list.
    pub contact_list_name: String,
}

/// Request to create a contact.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct CreateContactRequest {
    /// The name of the contact list.
    pub contact_list_name: String,
    /// The contact's email address.
    pub email_address: String,
    /// Topic preferences.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub topic_preferences: Option<Vec<super::TopicPreference>>,
    /// Unsubscribe all status.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub unsubscribe_all: Option<bool>,
    /// Custom attributes.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attributes_data: Option<String>,
}

impl CreateContactRequest {
    /// Create a new contact request.
    pub fn new(contact_list_name: impl Into<String>, email: impl Into<String>) -> Self {
        Self {
            contact_list_name: contact_list_name.into(),
            email_address: email.into(),
            topic_preferences: None,
            unsubscribe_all: None,
            attributes_data: None,
        }
    }
}

/// Request to delete a contact.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct DeleteContactRequest {
    /// The name of the contact list.
    pub contact_list_name: String,
    /// The contact's email address.
    pub email_address: String,
}

/// Request to add destinations to suppression list.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct PutSuppressedDestinationRequest {
    /// The email address to suppress.
    pub email_address: String,
    /// The reason for suppression.
    pub reason: SuppressionListReason,
}

impl PutSuppressedDestinationRequest {
    /// Create a new suppressed destination request.
    pub fn new(email: impl Into<String>, reason: SuppressionListReason) -> Self {
        Self {
            email_address: email.into(),
            reason,
        }
    }
}

/// Request to remove destination from suppression list.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct DeleteSuppressedDestinationRequest {
    /// The email address to remove from suppression.
    pub email_address: String,
}

/// Request to create an event destination.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct CreateConfigurationSetEventDestinationRequest {
    /// The name of the configuration set.
    pub configuration_set_name: String,
    /// The event destination details.
    pub event_destination: EventDestination,
    /// The name of the event destination.
    pub event_destination_name: String,
}

/// Request to delete an event destination.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct DeleteConfigurationSetEventDestinationRequest {
    /// The name of the configuration set.
    pub configuration_set_name: String,
    /// The name of the event destination.
    pub event_destination_name: String,
}

/// Request to put email identity DKIM attributes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct PutEmailIdentityDkimAttributesRequest {
    /// The email identity.
    pub email_identity: String,
    /// Whether DKIM signing is enabled.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signing_enabled: Option<bool>,
}

/// Request to put email identity mail from attributes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct PutEmailIdentityMailFromAttributesRequest {
    /// The email identity.
    pub email_identity: String,
    /// The MAIL FROM domain.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mail_from_domain: Option<String>,
    /// Behavior on MX failure.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub behavior_on_mx_failure: Option<super::BehaviorOnMxFailure>,
}

/// Request to put email identity feedback attributes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct PutEmailIdentityFeedbackAttributesRequest {
    /// The email identity.
    pub email_identity: String,
    /// Whether to forward bounce and complaint notifications.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email_forwarding_enabled: Option<bool>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{Content, Destination as EmailDest};

    #[test]
    fn test_send_email_request() {
        let content = EmailContent::new("Test Subject")
            .with_text("Test body");

        let request = SendEmailRequest::new(content)
            .with_from("sender@example.com")
            .with_configuration_set("my-config-set")
            .add_tag("campaign", "newsletter");

        assert_eq!(request.from_email_address, Some("sender@example.com".to_string()));
        assert_eq!(request.configuration_set_name, Some("my-config-set".to_string()));
        assert_eq!(request.email_tags.as_ref().unwrap().len(), 1);
    }

    #[test]
    fn test_create_email_identity_request() {
        let request = CreateEmailIdentityRequest::new("example.com")
            .add_tag("Environment", "Production")
            .with_configuration_set("default");

        assert_eq!(request.email_identity, "example.com");
        assert_eq!(request.tags.as_ref().unwrap().len(), 1);
    }

    #[test]
    fn test_template_content() {
        let content = TemplateContent::new()
            .with_subject("Welcome {{name}}")
            .with_html("<h1>Hello {{name}}</h1>")
            .with_text("Hello {{name}}");

        assert_eq!(content.subject, Some("Welcome {{name}}".to_string()));
        assert!(content.html.is_some());
        assert!(content.text.is_some());
    }
}
